### ISSUE TITLE
`torch.rfft` and `torch.ifft` arguments

### ISSUE THREAD

**Issue by scaomath** [zongyi-li/fourier_neural_operator#33]
@zongyi-li 

It seems that Torch has modified the `fft` to a namespace and put all functions under it: `rfft` and `ifft`, etc.

But right now there is no arguments for `one-sided`, and I checked torch's official document, now it seems everything is automatically one-sided unless you want specifically the negative part: [https://pytorch.org/docs/stable/fft.html](https://pytorch.org/docs/stable/fft.html). `normalized` arg became `norm`.

Can you please confirm?

**Comment by zongyi-li** [comment]
Torch 1.8 starts to support complex numbers. The old format has two dimensions at the end for the real and imaginary part, while the new using the `torch.cfloat` data type. As a result, they have a new version of `fft`. 

The new `torch.rfft` is automatically one-sided, and `normalized` becomes `norm`.

There are several changes that need to be made:

- Switch from `rfft`, `irfft` to `rfftn`, `irfftn`.
- Initialize the weight parameter as `torch.cfloat` instead of having the two dimensions at the end.
- Simply the complex multiplication for the new format.

The new version is about 30% faster. We plan to release it next week. If you want I can also send it to you sooner.

**Comment by scaomath** [comment]
Thanks. I will wait.

**Comment by zongyi-li** [comment]
Sorry it takes a bit longer to clean the code. But you may find this useful:
```
def compl_mul2d(a, b):
    # (batch, in_channel, x,y,t ), (in_channel, out_channel, x,y,t) -> (batch, out_channel, x,y,t)
    op = partial(torch.einsum, "bixy,ioxy->boxy")
    return op(a, b)

class SpectralConv2d(nn.Module):
    def __init__(self, in_channels, out_channels, modes1, modes2):
        super(SpectralConv2d, self).__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.modes1 = modes1 #Number of Fourier modes to multiply, at most floor(N/2) + 1
        self.modes2 = modes2

        self.scale = (1 / (in_channels * out_channels))
        self.weights1 = nn.Parameter(self.scale * torch.rand(in_channels, out_channels, self.modes1, self.modes2, dtype=torch.cfloat))
        self.weights2 = nn.Parameter(self.scale * torch.rand(in_channels, out_channels, self.modes1, self.modes2, dtype=torch.cfloat))

    def forward(self, x, size=None):
        if size==None:
            size = x.size(-1)

        batchsize = x.shape[0]
        #Compute Fourier coeffcients up to factor of e^(- something constant)
        x_ft = torch.fft.rfftn(x, dim=[2,3])

        # Multiply relevant Fourier modes
        out_ft = torch.zeros(batchsize, self.out_channels, size, size//2 + 1, device=x.device, dtype=torch.cfloat)
        out_ft[:, :, :self.modes1, :self.modes2] = \
            compl_mul2d(x_ft[:, :, :self.modes1, :self.modes2], self.weights1)
        out_ft[:, :, -self.modes1:, :self.modes2] = \
            compl_mul2d(x_ft[:, :, -self.modes1:, :self.modes2], self.weights2)


        #Return to physical space
        x = torch.fft.irfftn(out_ft, s=(size, size), dim=[2,3])
        return x
```

**Comment by scaomath** [comment]
This is my current wrapper and spectral conv code based on yours. It uses the old code and not the `cfloat`I will start a pull request. If you think this looks good, I can clean a bit and start a PR.
I changed the function name to have `matmul` string in it since I have a debugger checking the `__name__`, `__qualname__` of the functions.

```python
class SpectralConv2d(nn.Module):
    def __init__(self, in_dim, # number of hidden features out from attention to the fourier conv
                       out_dim, # number of hidden features to the final regressor
                       modes: int, # number of fourier modes
                       n_grid=None,
                       dropout=0.1,
                       debug=False):
        super(SpectralConv2d, self).__init__()

        '''
        Modified Zongyi Li's SpectralConv2d code
        '''
        self.in_dim = in_dim
        self.out_dim = out_dim
        self.linear = nn.Linear(in_dim, out_dim) # for residual
        self.modes = modes
        self.activation = nn.SiLU()
        self.n_grid = n_grid # just for debugging
        fourier_weight = [Parameter(torch.FloatTensor(in_dim, out_dim, modes, modes, 2)) for _ in range(2)]
        self.fourier_weight = nn.ParameterList(fourier_weight)
        for param in self.fourier_weight:
            xavier_normal_(param, gain=1/(in_dim*out_dim))
        self.dropout = nn.Dropout(dropout)
        self.debug = debug

    @staticmethod
    def complex_matmul_2d(a, b):
        # (batch, in_channel, x, y), (in_channel, out_channel, x, y) -> (batch, out_channel, x, y)
        op = partial(torch.einsum, "bixy,ioxy->boxy")
        return torch.stack([
            op(a[..., 0], b[..., 0]) - op(a[..., 1], b[..., 1]),
            op(a[..., 1], b[..., 0]) + op(a[..., 0], b[..., 1])
        ], dim=-1)

    def forward(self, x):
        '''
        Input: (-1, n_grid**2, in_features) or (-1, n_grid, n_grid, in_features)
        Output: (-1, n_grid**2, out_features) or (-1, n_grid, n_grid, out_features)
        '''
        batch_size = x.size(0)
        n_dim = x.ndim
        if n_dim == 4:
            n = x.size(1)
            assert x.size(1) == x.size(2)
        elif n_dim == 3:
            n = int(x.size(1)**(0.5))
        else:
            raise ValueError("Dimension not implemented")
        in_dim = self.in_dim
        out_dim = self.out_dim
        modes = self.modes

        if self.debug:
            print(f"grid size: \t {(n, self.n_grid)}")
            print(f"x size: \t {x.size()}")
            assert n == self.n_grid
        
        x = x.view(-1, n, n, in_dim) # no need to use contiguous here. 
        res = self.linear(x) #(B, n, n, E_out)
        x = self.dropout(x)


        if self.debug:
            print(f"signal in space: \t { x.size()}")
        x = x.permute(0, 3, 1, 2) # (B, n, n, E) -> # (B, E, n, n)
        # f: Nyquist frequency
        x_ft = fft.rfft2(x, s=(n, n), norm="ortho")  # (B, E, n, f)
        x_ft = torch.stack([x_ft.real, x_ft.imag], dim=4) # (B, E, n, f, 2)
        if self.debug:
            print(f"signal in freq: \t {x_ft.size()}")

        # out_ft second dim should be out_dim, this is a bug in original Zongyi Li's code
        out_ft = torch.zeros(batch_size, out_dim, n, n//2+1, 2, device=x.device)
        out_ft[:, :, :modes, :modes] = self.complex_matmul_2d(x_ft[:, :, :modes, :modes], self.fourier_weight[0])
        out_ft[:, :, -modes:, :modes] = self.complex_matmul_2d(x_ft[:, :, -modes:, :modes], self.fourier_weight[1])
        out_ft = torch.complex(out_ft[...,0], out_ft[...,1])
        if self.debug:
            print(f"regression in freq: \t {out_ft.size()}")

        x = fft.irfft2(out_ft, s=(n, n), norm="ortho") #(B, E_out, n, n)
        x = x.permute(0, 2, 3, 1) #(B, n, n, E_out)
        x = self.activation(x + res)

        if n_dim == 3:
            x = x.view(batch_size, n**2, out_dim) 

        if self.debug:
            print(f"out in space: \t {x.size()}")

        return x


class SpectralRegressor2d(nn.Module):
    def __init__(self, in_dim, # number of hidden features out from attention to the fourier conv
                       freq_dim, # number of frequency features
                       out_dim, # number of target dim
                       modes: int, # number of fourier modes
                       num_spectral_layers:int = 2,
                       n_grid=None,
                       dim_feedforward=None,
                       spacial_fc=False,
                       spacial_dim=3,
                       dropout=0.1,
                       debug=False):
        super(SpectralRegressor2d, self).__init__()
        '''
        A wrapper for SpectralConv2d
        '''
        self.spacial_fc = spacial_fc
        self.fc = nn.Linear(spacial_dim, in_dim) if self.spacial_fc else None
        self.spectral_conv = nn.ModuleList([SpectralConv2d(in_dim=in_dim, 
                                                           out_dim=freq_dim, 
                                                           n_grid=n_grid,
                                                           modes=modes,
                                                           dropout=dropout,
                                                           debug=debug)])
        for _ in range(num_spectral_layers - 1):
            self.spectral_conv.append(SpectralConv2d(in_dim=freq_dim, 
                                                     out_dim=freq_dim, 
                                                     n_grid=n_grid,
                                                     modes=modes,
                                                     dropout=dropout,
                                                     debug=debug))

        self.n_grid = n_grid # dummy 
        self.dim_feedforward = 2*freq_dim if dim_feedforward is None else dim_feedforward
        self.regressor = nn.Sequential(
                                    nn.Linear(freq_dim, self.dim_feedforward), 
                                    nn.SiLU(),
                                    nn.Linear(self.dim_feedforward, out_dim),
                                )

    def forward(self, x, pos=None):
        '''
        Input: (-1, n*n, in_features)
        Output: (-1, n*n, n_targets)
        '''
        if self.spacial_fc:
            x = torch.cat([x, pos], dim=-1)
            x = self.fc(x)

        for layer in self.spectral_conv:
            x = layer(x)

        x = self.regressor(x)
        return x
```

**Comment by scaomath** [comment]
This is my current wrapper and spectral conv code based on yours. It uses the old code and not the `cfloat`. If you think this looks good, I can clean a bit and start a PR.
I changed the function name to have `matmul` string in it since I have a debugger checking the `__name__`, `__qualname__` of the functions. 

I will give a casual talk this Thursday related to your work. Feel free to join: https://math.wustl.edu/events/statistics-and-data-science-seminar-transformer-dissection-amateur-applied-mathematician?d=2021-04-15 

```python
class SpectralConv2d(nn.Module):
    def __init__(self, in_dim, # number of hidden features to the fourier conv
                       out_dim, # number of hidden features to the final regressor
                       modes: int, # number of fourier modes
                       n_grid=None,
                       dropout=0.1,
                       debug=False):
        super(SpectralConv2d, self).__init__()

        '''
        Modified Zongyi Li's SpectralConv2d code
        '''
        self.in_dim = in_dim
        self.out_dim = out_dim
        self.linear = nn.Linear(in_dim, out_dim) # for residual
        self.modes = modes
        self.activation = nn.SiLU()
        self.n_grid = n_grid # just for debugging
        fourier_weight = [Parameter(torch.FloatTensor(in_dim, out_dim, modes, modes, 2)) for _ in range(2)]
        self.fourier_weight = nn.ParameterList(fourier_weight)
        for param in self.fourier_weight:
            xavier_normal_(param, gain=1/(in_dim*out_dim))
        self.dropout = nn.Dropout(dropout)
        self.debug = debug

    @staticmethod
    def complex_matmul_2d(a, b):
        # (batch, in_channel, x, y), (in_channel, out_channel, x, y) -> (batch, out_channel, x, y)
        op = partial(torch.einsum, "bixy,ioxy->boxy")
        return torch.stack([
            op(a[..., 0], b[..., 0]) - op(a[..., 1], b[..., 1]),
            op(a[..., 1], b[..., 0]) + op(a[..., 0], b[..., 1])
        ], dim=-1)

    def forward(self, x):
        '''
        Input: (-1, n_grid**2, in_features) or (-1, n_grid, n_grid, in_features)
        Output: (-1, n_grid**2, out_features) or (-1, n_grid, n_grid, out_features)
        '''
        batch_size = x.size(0)
        n_dim = x.ndim
        if n_dim == 4:
            n = x.size(1)
            assert x.size(1) == x.size(2)
        elif n_dim == 3:
            n = int(x.size(1)**(0.5))
        else:
            raise ValueError("Dimension not implemented")
        in_dim = self.in_dim
        out_dim = self.out_dim
        modes = self.modes

        if self.debug:
            print(f"grid size: \t {(n, self.n_grid)}")
            print(f"x size: \t {x.size()}")
            assert n == self.n_grid
        
        x = x.view(-1, n, n, in_dim) # no need to use contiguous here. 
        res = self.linear(x) #(B, n, n, E_out)
        x = self.dropout(x)


        if self.debug:
            print(f"signal in space: \t { x.size()}")
        x = x.permute(0, 3, 1, 2) # (B, n, n, E) -> # (B, E, n, n)
        # f: Nyquist frequency
        x_ft = fft.rfft2(x, s=(n, n), norm="ortho")  # (B, E, n, f)
        x_ft = torch.stack([x_ft.real, x_ft.imag], dim=4) # (B, E, n, f, 2)
        if self.debug:
            print(f"signal in freq: \t {x_ft.size()}")

        # out_ft second dim should be out_dim, this is a bug in original Zongyi Li's code
        out_ft = torch.zeros(batch_size, out_dim, n, n//2+1, 2, device=x.device)
        out_ft[:, :, :modes, :modes] = self.complex_matmul_2d(x_ft[:, :, :modes, :modes], self.fourier_weight[0])
        out_ft[:, :, -modes:, :modes] = self.complex_matmul_2d(x_ft[:, :, -modes:, :modes], self.fourier_weight[1])
        out_ft = torch.complex(out_ft[...,0], out_ft[...,1])
        if self.debug:
            print(f"regression in freq: \t {out_ft.size()}")

        x = fft.irfft2(out_ft, s=(n, n), norm="ortho") #(B, E_out, n, n)
        x = x.permute(0, 2, 3, 1) #(B, n, n, E_out)
        x = self.activation(x + res)

        if n_dim == 3:
            x = x.view(batch_size, n**2, out_dim) 

        if self.debug:
            print(f"out in space: \t {x.size()}")

        return x


class SpectralRegressor2d(nn.Module):
    def __init__(self, in_dim, # number of hidden features to the fourier conv
                       freq_dim, # number of frequency features
                       out_dim, # number of target dim
                       modes: int, # number of fourier modes
                       num_spectral_layers:int = 2,
                       n_grid=None,
                       dim_feedforward=None,
                       spacial_fc=False,
                       spacial_dim=3,
                       dropout=0.1,
                       debug=False):
        super(SpectralRegressor2d, self).__init__()
        '''
        A wrapper for SpectralConv2d
        '''
        self.spacial_fc = spacial_fc
        self.fc = nn.Linear(spacial_dim, in_dim) if self.spacial_fc else None
        self.spectral_conv = nn.ModuleList([SpectralConv2d(in_dim=in_dim, 
                                                           out_dim=freq_dim, 
                                                           n_grid=n_grid,
                                                           modes=modes,
                                                           dropout=dropout,
                                                           debug=debug)])
        for _ in range(num_spectral_layers - 1):
            self.spectral_conv.append(SpectralConv2d(in_dim=freq_dim, 
                                                     out_dim=freq_dim, 
                                                     n_grid=n_grid,
                                                     modes=modes,
                                                     dropout=dropout,
                                                     debug=debug))

        self.n_grid = n_grid # dummy 
        self.dim_feedforward = 2*freq_dim if dim_feedforward is None else dim_feedforward
        self.regressor = nn.Sequential(
                                    nn.Linear(freq_dim, self.dim_feedforward), 
                                    nn.SiLU(),
                                    nn.Linear(self.dim_feedforward, out_dim),
                                )

    def forward(self, x, pos=None):
        '''
        Input: (-1, n*n, in_features)
        Output: (-1, n*n, n_targets)
        '''
        if self.spacial_fc:
            x = torch.cat([x, pos], dim=-1)
            x = self.fc(x)

        for layer in self.spectral_conv:
            x = layer(x)

        x = self.regressor(x)
        return x
```

**Comment by zongyi-li** [comment]
Oh you are from WashU. I had my college there. I will be very happy to know your progress with our method.
The output of `fft.rfft2()` should be torch.cfloat type, does the old `complex_matmul_2d()` still work? I would suggest you try the newer implementation given above.

**Comment by scaomath** [comment]
Yeah, it still works. Concatenating the real and complex part in the last dim, and using the old code is good as essentially the original function you implemented is a complex multiplication `(A + i B)` multiply `(C + i D)` in every channel.
